control-services/cli/config.py

```diff
@@ -151,13 +151,41 @@ def __init__(self, jsonConfigFile, envPath):
             cfg['host']['vater_dir_path'] + cfg['vater_repo']['rel_image_path'] +           \
             'jenkins/'
 
+        cfg['jenkins']['data_dir_path'] =                                                   \
+            cfg['host']['vater_dir_path'] + 'control-services/data/jenkins/'
+
         cfg['jenkins']['casc_file_path'] = cfg['jenkins']['image_dir_path'] + 'casc.yaml'   \
 
         cfg['jenkins']['casc'] = {}
-        cfg['jenkins']['casc']['unclassified'] = {}
+
+        cfg['jenkins']['casc']['jenkins'] = {}        
```

Comment:
Most of this seems like it could be declared straight in the casc file - why are we manually declaring it inline in python?

Reply:
`config.json` -> `config.py` -> `casc.yaml`

Some of the information that lands in `casc.yaml` is configurable in `config.json`, and `config.py` will manipulate that data appropriately.

For instance, 
```
cfg['jenkins']['casc']['unclassified']['location']['url'] =                         \
            'http://' + cfg['host']['ip'] + '/' + cfg['jenkins']['port'] + '/'
```

`cfg['jenkins']['port']` comes from `config.json` and is subject to change.

Reply:
true, but much of it is hard-coded into this "config.py" file and is not pulled from `config.json` - in those cases, wouldn't it be preferable to have the key/value pairs declared directly in the casc file?

Reply:
For example,
```python
        cfg['jenkins']['casc']['jenkins']['securityRealm']['local']                         \
            ['allowsSignup'] = 'false
```

Reply:
Agreed. In your opinion:

1) How much more to we anticipate adding to the `casc.yaml` that could come from the `config.json`? If it's more than the one line, using multiple methods to build the configuration (static file and config.py) might be more difficult the maintain. 

2) Is there an elegant solution for altering pieces of a `casc.yaml` in place? Keeping the `dict` is very clear due to the named key/value format and it all happens in one file with little formatting due the `yaml` module. If we use a canary replacement method we have to open the file, alter the canary, and then rebuild the file. 

3) How do we handle consecutive runs with different configurations? Because we build and move the `casc.yaml` the file always reflect the most current changes. We could keep a static template file, read it in, alter the canary values, and write it back, but that's the same as keeping the configuration in dict form

4) Not a question, but a nice thing about the current format, because all configuration makes it's way into `config.py`, and the object is passed across different services, anything in the `casc.yaml` is available to any other service without additional processing. Do you ever anticipate the `casc.yaml` configuration being relevant to another service such that the additional complexity of re-locating the file and parsing is a detriment?

Reply:
I don't think a template/canary file would really solve the problem here. If we want to do consecutive runs with different configurations, to me the best solution seems like environment variables. A local copy of a .env can be read in when the docker container is run. HOWEVER, it's important to note that I think this is out of scope of your PR and the .env solution shouldn't be implemented in this PR. The only thing I was suggesting changing for now is the static values, and the dynamic values could be altered to be read in from the environment later.

So in answer to your questions:
1. I think it's valid to anticipate needing further `config.json` values in the future, but I'm also suggesting that this doesn't necessarily mean we need to build the configuration file in more than one place (per the solution suggested above)
2. I don't think we should alter the `casc.yaml` in place (see above)
3. Env variables are a good solution to this as well
4. I'm not sure why other services would need information from the `casc.yaml`, but if they did, changing it to an environment variable would be simple

Another possibility is running a script that converts the config to an in-RAM set of env variables at runtime.

Reply:
There already is a `.env` passed to `docker-compose`. The path comes from `parser.py` and `vDocker.py` implements the functions that interfaces with `docker-compose` -> `docker engine`.

```
control@control:~/vater/control-services$ cat .env
gitea_db_user=gitea
gitea_db_password=gitea
semaphore_admin_password=cangetin
```

However, this file is currently only used to pass on information from `config.json` or the total configuration built in `config.py`.

Are you suggesting keeping the `casc.yaml` static:

```
control@control:~/vater/control-services/images/jenkins$ cat casc.yaml
authorizationStrategy:
  globalMatrix:
    permissions:
    - Overall/Administer:admin
    - Overall/Read:authenticated
security:
  queueItemAuthenticator:
    authenticators:
    - global:
        strategy: triggeringUsersAuthorizationStrategy
unclassified:
  location:
    url: http://192.168.100.1/5000/
```

But passing the `unclassified` block to the `docker-compose up  jenkins` via `.env` instead, because it's dynamic / can change per environment?

```
unclassified:
  location:
    url: http://192.168.100.1/5000
```

If so, what would the `.env` look like to get that into one file with the static `casc.yaml`?

Also, the current `.env` is also build in `config.py`:

```
228         # Docker variables
229         cfg['docker'] = {}
230
231         cfg['docker']['compose_file_path'] =                                                \
232             cfg['host']['vater_dir_path'] + 'control-services/docker-compose.yml'
233
234         cfg['docker']['env_path'] = envPath
235
236         cfg['docker']['env'] = []
237
238         cfg['docker']['env'].append(
239             'gitea_db_user=' + cfg['gitea_db']['user']
240         )
241
242         cfg['docker']['env'].append(
243             'gitea_db_password=' + cfg['gitea_db']['password']
244         )
245
246         cfg['docker']['env'].append(
247             'semaphore_admin_password=' + cfg['semaphore']['password']
248         )
```

Reply:
@marissaeinhorn @rylagek thoughts?

Reply:
For example:

```yaml
unclassified:
  location:
    url: "http://${HOST_IP}:${JENKINS_PORT}/"
```

Then add `HOST_IP` and `JENKINS_PORT` (or whatever you want to call it) to the `.env` and `docker-compose.yml`

Reply:
I have no qualms with the code as is but I am open to the alterations proposed by @rmfirth